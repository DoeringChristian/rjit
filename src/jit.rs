use std::collections::HashMap;
use std::fmt::Write;
use std::ops::Range;
use std::sync::Arc;

use anyhow::{anyhow, Result};
use itertools::Itertools;

use crate::backend::Kernel;
use crate::schedule::{Env, ScheduleIr};
use crate::trace::Internal;
use crate::var::{Data, Op};

// TODO: pooling for paralel execution
///
/// The Jit Compiler first generates schedules (Intermediate Representation) from a Trace.
/// It then assembles and compiles a Kernel depending on the Backend.
///
/// Ir -> [Schedule; N] -> [Kernel; N]
///
/// Where N is the number of schedule groups.
/// These are extracted from the scheduled variables in the Ir.
///
#[derive(Debug, Default)]
pub struct Jit {
    pub kernels: HashMap<KernelKey, Arc<dyn Kernel>>,
    pub kernel_history: Vec<KernelHistroyEntry>,
}

#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Clone)]
pub enum KernelKey {
    Hash(u128),
    Name(String),
}
impl std::fmt::Display for KernelKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            KernelKey::Hash(hash) => write!(f, "{hash}"),
            KernelKey::Name(name) => write!(f, "{name}"),
        }
    }
}

#[derive(Debug)]
pub struct KernelHistroyEntry {
    key: KernelKey,
    size: usize,
    backend: &'static str,
}

#[derive(Debug)]
struct ScheduleGroup {
    size: usize,
    range: Range<usize>,
}

impl Jit {
    ///
    /// Evaluates a Ir by first constructing Schedules, which are then compiled and assembled
    /// into kernels.
    ///
    /// A the end, all scheduled variables are overwritten with the calculated data.
    ///
    pub fn eval(&mut self, ir: &mut Internal) -> Result<()> {
        if ir.scheduled.len() == 0 {
            log::trace!("Nothing to evaluate.");
            return Ok(());
        }
        let mut schedule = ir.scheduled.clone();

        // For every scheduled variable (destination) we have to create a new buffer (except if it
        // is void)
        for id in ir.scheduled.clone() {
            let var = ir.var(id);
            // Do not reallocate on scheduled variables (don't yet know if this is right)
            // TODO: better test
            if !var.data.is_buffer() && var.ty.size() > 0 {
                let size = ir.var(id).size;
                let ty_size = ir.var(id).ty.size();
                let buffer = ir
                    .backend
                    .as_ref()
                    .ok_or(anyhow!("Backend not initialized!"))?
                    .buffer_uninit(size * ty_size)?;

                let mut var = ir.var_mut(id);
                var.data = Data::Buffer(buffer);
            }
        }

        schedule.sort_by(|id0, id1| ir.var(*id0).size.cmp(&ir.var(*id1).size));
        let mut schedule_groups = vec![];

        let mut current = 0;
        for i in 1..schedule.len() {
            let size = ir.var(schedule[i - 1]).size;
            if size != ir.var(schedule[i]).size {
                schedule_groups.push(ScheduleGroup {
                    size,
                    range: current..i,
                });
                current = i;
            }
        }
        schedule_groups.push(ScheduleGroup {
            size: ir.var(schedule[current]).size,
            range: current..schedule.len(),
        });

        let futures = schedule_groups
            .into_iter()
            .map(|sg| {
                let mut s = ScheduleIr::new();
                let mut env = Env::default();
                s.collect_vars(&mut env, ir, &schedule[sg.range]);
                let hash = s.internal_hash();
                let key = KernelKey::Hash(hash);

                let kernel = self.kernels.entry(key.clone()).or_insert(
                    ir.backend
                        .as_ref()
                        .ok_or(anyhow!("Backend not initialized!"))?
                        .compile_kernel(&s, &env)?,
                );

                self.kernel_history.push(KernelHistroyEntry {
                    key,
                    size: sg.size,
                    backend: kernel.backend_ident(),
                });

                kernel.execute_async(&mut env, sg.size)
            })
            .collect::<Vec<_>>();

        // TODO: synchronisation here
        for future in futures {
            future?.wait();
        }
        // ir.backend.as_ref().unwrap().synchronize();

        // After executing the kernels, the Ir is cleaned up.
        // To do so, we first decrement the refcount and then set the ParamType to Input and op to
        // Data
        for id in ir.scheduled.clone() {
            let var = ir.var_mut(id);

            // Set op and type for next kernel:
            var.op = Op::Data;
            var.dirty = false;

            // Clear dependencies:
            let deps = var.deps.clone();
            var.deps.clear();

            for dep in deps {
                ir.dec_rc(dep);
            }

            ir.dec_rc(id);
        }
        ir.scheduled.clear();
        Ok(())
    }
    // pub fn launch_kernel(&self, ir: &mut Internal) {}
    ///
    /// Writes the kernel assemblies into a string which can then be checked by snapshot testing
    /// tools such as insta.
    ///
    pub fn kernel_history(&self) -> String {
        let mut s = String::new();
        writeln!(s, "Kernel History:").unwrap();
        for entry in self.kernel_history.iter() {
            writeln!(
                s,
                "Launched {} Kernel {} with {} elements",
                entry.backend, entry.key, entry.size
            )
            .unwrap();
        }

        for (hash, kernel) in self
            .kernels
            .iter()
            .sorted_by(|(hash1, _), (hash2, _)| hash1.cmp(hash2))
        {
            writeln!(s, "===============================================").unwrap();
            writeln!(s, "Kernel {}:", hash).unwrap();
            writeln!(s, "").unwrap();
            write!(s, "{}", kernel.assembly()).unwrap();
        }
        s
    }
}

#[cfg(test)]
mod test {}
