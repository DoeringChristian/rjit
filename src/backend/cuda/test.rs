use crate::jit::Jit;
use crate::trace::{ReduceOp, Trace};
use anyhow::Result;

#[test]
fn refcounting() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[1f32; 10])?;
    assert_eq!(x.var().rc, 1, "rc of x should be 1 (in x)");
    let y = x.add(&x)?;
    // let y = ir::add(&x, &x);

    assert_eq!(
        x.var().rc,
        3,
        "rc of x should be 3 (2 refs in y and 1 in x)"
    );

    assert_eq!(y.var().rc, 1, "rc of y should be 1 (in y)");

    ir.schedule(&[&y]);

    assert_eq!(
        y.var().rc,
        2,
        "rc of y should be 2 (1 in y and 1 in schedule)"
    );

    ir.eval()?;

    assert_eq!(
        x.var().rc,
        1,
        "rc of x should be 1 (dependencies of y shuld be cleaned)"
    );
    assert_eq!(
        y.var().rc,
        1,
        "rc of y should be 2 (y from schedule should be cleaned)"
    );

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![2f32; 10]);
    Ok(())
}
#[test]
fn load_add_f32() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[1f32; 10])?;
    // let y = ir::add(&x, &x);
    let y = x.add(&x)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![2f32; 10]);
    Ok(())
}
#[test]
fn load_gather_f32() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[1f32, 2., 3., 4., 5.])?;
    let i = ir.array(&[0u32, 1, 4])?;
    let y = x.gather(&i, None)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![1., 2., 5.]);
    Ok(())
}
#[test]
fn reindex() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.index(10);

    let i = ir.index(3);
    let c = ir.literal(2u32)?;
    let i = i.add(&c)?;

    let y = x.gather(&i, None)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<u32>().unwrap(), vec![2, 3, 4]);
    Ok(())
}
#[test]
fn index() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(&["cuda"])?;

    let i = ir.index(10);

    ir.schedule(&[&i]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(
        i.to_host::<u32>().unwrap(),
        vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    );
    Ok(())
}
#[test]
fn gather_eval() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let r = {
        let x = ir.index(3);
        let y = ir.array(&[1u32, 2, 3])?;

        // let z = ir::add(&x, &y);
        let z = x.add(&y)?;

        let r = z.gather(&ir.index(3), None)?;
        r
    };
    // schedule

    ir.schedule(&[&r]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(r.to_host::<u32>().unwrap(), vec![1, 3, 5]);
    Ok(())
}
#[test]
fn paralell() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.index(10);

    let y = ir.index(3);

    ir.schedule(&[&x, &y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(
        x.to_host::<u32>().unwrap(),
        vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    );
    assert_eq!(y.to_host::<u32>().unwrap(), vec![0, 1, 2]);
    Ok(())
}
#[test]
fn _load_gather() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[1f32, 2., 3.])?;

    ir.schedule(&[&x]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<f32>().unwrap(), vec![1., 2., 3.]);
    Ok(())
}
#[test]
fn eval_scatter() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;
    let c = ir.literal(1u32)?;
    let x = x.add(&c)?; // x: [1, 1, 1, 1]

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?; // i: [1, 2, 3]

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?; // x: [1, 2, 2, 2]

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![1, 2, 2, 2]);
    Ok(())
}
#[test]
fn scatter_twice() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?;

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?;

    let i = ir.index(2);

    let y = ir.literal(3u32)?;

    y.scatter(&x, &i, None)?;

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![3, 3, 2, 2]);
    Ok(())
}
#[test]
fn scatter_twice_add() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?;

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?;

    let i = ir.index(2);

    let y = ir.literal(3u32)?;

    y.scatter(&x, &i, None)?;

    let c = ir.literal(1u32)?;
    let x = x.add(&c)?;

    ir.schedule(&[&x]);

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![4, 4, 3, 3]);
    Ok(())
}
#[test]
fn scatter_reduce() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.array(&[0u32, 0, 0])?;

    let y = ir.literal(1u32)?;

    y.scatter_reduce(&x, &i, None, ReduceOp::Add)?;

    ir.schedule(&[&x]);

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![3, 0, 0, 0]);
    Ok(())
}
#[test]
fn tex_lookup() -> Result<()> {
    pretty_env_logger::try_init().ok();
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0.5f32])?;
    let y = ir.array(&[0.5f32])?;

    let data = ir.array(&[1.0f32; 400])?;

    let tex = data.to_texture(&[10, 10], 4)?;

    let res = tex.tex_lookup(&[&x, &y])?;

    let r = res[0].clone();

    r.schedule();

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(r.to_host::<f32>().unwrap(), vec![1.]);
    Ok(())
}
#[test]
fn cast() -> Result<()> {
    pretty_env_logger::try_init().ok();
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[1u32, 2, 3])?;
    let x = x.cast(&crate::VarType::U64)?;

    x.schedule();

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u64>().unwrap(), vec![1, 2, 3]);
    Ok(())
}
#[test]
fn bitcast() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.array(&[0x3f800000u32, 0x40000000, 0x40400000])?;

    let x = x.bitcast(&crate::VarType::F32)?;

    x.schedule();
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<f32>().unwrap(), vec![1., 2., 3.]);
    Ok(())
}
#[test]
fn scatter_const_mask() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.literal(1f32)?;
    let mask = ir.array(&[true, false, true])?;

    let dst = ir.array(&[0f32, 0., 0.])?;

    x.scatter(&dst, &ir.index(3), Some(&mask))?;

    ir.eval()?;
    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(dst.to_host::<f32>().unwrap(), vec![1., 0., 1.]);
    Ok(())
}
#[test]
fn scatter_gather() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let dst = ir.array(&[0u32, 0, 0])?;
    let idx = ir.index(3);

    let x = ir.array(&[1u32, 1, 1])?;

    x.scatter(&dst, &idx, None)?;

    let y = dst.gather(&idx, None)?;

    y.schedule();
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(dst.to_host::<u32>().unwrap(), vec![1, 1, 1]);
    assert_eq!(y.to_host::<u32>().unwrap(), vec![1, 1, 1]);
    Ok(())
}
