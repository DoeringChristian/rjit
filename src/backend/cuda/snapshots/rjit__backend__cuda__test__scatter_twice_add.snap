---
source: src/backend/cuda/test.rs
expression: ir.kernel_history()
---
Kernel History:
Launched CUDA Kernel 231069604177812995 with 3 elements
Launched CUDA Kernel 4585386102060764057 with 2 elements
Launched CUDA Kernel 2654645005488410590 with 4 elements
===============================================
Kernel 231069604177812995:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[24]) {

	.reg.b8   %b <11>; .reg.b16 %w<11>; .reg.b32 %r<11>;
	.reg.b64  %rd<11>; .reg.f32 %f<11>; .reg.f64 %d<11>;
	.reg.pred %p <11>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 5, buf: None, tex: None, accel: None, opaque: None, literal: 2, size: 1, sbt_hash: 0 } =>
	mov.b32 %r5, 0x2;


	// [1]: ScheduleVar { op: Data, deps: [], ty: U32, param_ty: None, reg: 6, buf: Some(0), tex: None, accel: None, opaque: None, literal: 0, size: 0, sbt_hash: 0 } =>

	// [2]: ScheduleVar { op: Idx, deps: [], ty: U32, param_ty: None, reg: 8, buf: None, tex: None, accel: None, opaque: None, literal: 0, size: 3, sbt_hash: 0 } =>
	mov.u32 %r8, %r0;


	// [3]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 9, buf: None, tex: None, accel: None, opaque: None, literal: 1, size: 1, sbt_hash: 0 } =>
	mov.b32 %r9, 0x1;


	// [4]: ScheduleVar { op: Add, deps: [SVarId(2), SVarId(3)], ty: U32, param_ty: None, reg: 7, buf: None, tex: None, accel: None, opaque: None, literal: 0, size: 3, sbt_hash: 0 } =>
	add.u32 %r7, %r8, %r9;

	// [5]: ScheduleVar { op: Literal, deps: [], ty: Bool, param_ty: None, reg: 10, buf: None, tex: None, accel: None, opaque: None, literal: 1, size: 1, sbt_hash: 0 } =>
	mov.pred %p10, 0x1;


	// [6]: ScheduleVar { op: Scatter { op: None }, deps: [SVarId(0), SVarId(1), SVarId(4), SVarId(5)], ty: Void, param_ty: None, reg: 4, buf: None, tex: None, accel: None, opaque: None, literal: 0, size: 3, sbt_hash: 0 } =>
	@!%p10 bra l_4_done;

	ld.param.u64 %rd0, [params+8];
	mad.wide.u32 %rd3, %r7, 4, %rd0;
	st.global.u32 [%rd3], %r5;
	l_4_done:

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}
===============================================
Kernel 2654645005488410590:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[32]) {

	.reg.b8   %b <7>; .reg.b16 %w<7>; .reg.b32 %r<7>;
	.reg.b64  %rd<7>; .reg.f32 %f<7>; .reg.f64 %d<7>;
	.reg.pred %p <7>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Data, deps: [], ty: U32, param_ty: Input, reg: 5, buf: Some(0), tex: None, accel: None, opaque: None, literal: 0, size: 4, sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+8];
	mad.wide.u32 %rd0, %r0, 4, %rd0;
	ld.global.cs.u32 %r5, [%rd0];

	// [1]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 6, buf: None, tex: None, accel: None, opaque: None, literal: 1, size: 1, sbt_hash: 0 } =>
	mov.b32 %r6, 0x1;


	// [2]: ScheduleVar { op: Add, deps: [SVarId(0), SVarId(1)], ty: U32, param_ty: Output, reg: 4, buf: Some(1), tex: None, accel: None, opaque: None, literal: 0, size: 4, sbt_hash: 0 } =>
	add.u32 %r4, %r5, %r6;

	// Store:
	ld.param.u64 %rd0, [params + 16]; // rd0 <- params[offset]
	mad.wide.u32 %rd0, %r0, 4, %rd0; // rd0 <- Index * ty.size() + params[offset]
	st.global.cs.u32 [%rd0], %r4; // (Index * ty.size() + params[offset])[Index] <- var

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}
===============================================
Kernel 4585386102060764057:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[24]) {

	.reg.b8   %b <9>; .reg.b16 %w<9>; .reg.b32 %r<9>;
	.reg.b64  %rd<9>; .reg.f32 %f<9>; .reg.f64 %d<9>;
	.reg.pred %p <9>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 5, buf: None, tex: None, accel: None, opaque: None, literal: 3, size: 1, sbt_hash: 0 } =>
	mov.b32 %r5, 0x3;


	// [1]: ScheduleVar { op: Data, deps: [], ty: U32, param_ty: None, reg: 6, buf: Some(0), tex: None, accel: None, opaque: None, literal: 0, size: 0, sbt_hash: 0 } =>

	// [2]: ScheduleVar { op: Idx, deps: [], ty: U32, param_ty: None, reg: 7, buf: None, tex: None, accel: None, opaque: None, literal: 0, size: 2, sbt_hash: 0 } =>
	mov.u32 %r7, %r0;


	// [3]: ScheduleVar { op: Literal, deps: [], ty: Bool, param_ty: None, reg: 8, buf: None, tex: None, accel: None, opaque: None, literal: 1, size: 1, sbt_hash: 0 } =>
	mov.pred %p8, 0x1;


	// [4]: ScheduleVar { op: Scatter { op: None }, deps: [SVarId(0), SVarId(1), SVarId(2), SVarId(3)], ty: Void, param_ty: None, reg: 4, buf: None, tex: None, accel: None, opaque: None, literal: 0, size: 2, sbt_hash: 0 } =>
	@!%p8 bra l_4_done;

	ld.param.u64 %rd0, [params+8];
	mad.wide.u32 %rd3, %r7, 4, %rd0;
	st.global.u32 [%rd3], %r5;
	l_4_done:

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}

